### PR TITLE
feat(server): add cache module

### PR DESCRIPTION
**Pull request by forehalo**

### CODE REVIEW

**Comment by forehalo** [comment]
Current dependencies on/for this PR:
* canary
  * **PR #4970** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/4970?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
    * **PR #4971** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/4971?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
      * **PR #5002** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/5002?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
        * **PR #5004** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/5004?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
      * **PR #4973** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/4973?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈
        * **PR #4975** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/4975?utm_source=stack-comment-icon" target="_blank">Graphite</a> 

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/toeverything/AFFiNE/4973?utm_source=stack-comment).

**Review by EYHN — DISMISSED** [review]

File: packages/backend/server/src/cache/index.ts [hunk 1]
```diff
@@ -0,0 +1,23 @@
+import { FactoryProvider, Global, Module } from '@nestjs/common';
+import { Redis as Cache } from 'ioredis';
+
+import { Config } from '../config';
+
+const RedisProvider: FactoryProvider = {
+  provide: Cache,
+  useFactory: (config: Config) => {
+    if (config.redis.enabled) {
+      new Cache(config.redis);
```
Comment (EYHN) [inline packages/backend/server/src/cache/index.ts:10]: return ?

File: packages/backend/server/src/cache/redis.ts [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+
+  async mapDecrease(
+    map: string,
+    key: string,
+    count: number = 1
+  ): Promise<number> {
+    return this.redis.hincrby(map, key, -count);
+  }
+
+  async mapGet(map: string, key: string): Promise<any> {
```
Comment (Brooooooklyn) [inline packages/backend/server/src/cache/redis.ts]:
```suggestion
  async mapGet<T = any>(map: string, key: string): Promise<T | undefined> {
```

File: packages/backend/server/src/cache/redis.ts [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+  }
+
+  async popFront(key: string, count: number = 1): Promise<any[]> {
+    return this.redis
+      .lpop(key, count)
+      .then(data => (data ?? []).map(v => JSON.parse(v)))
+      .catch(() => []);
+  }
+
+  async popBack(key: string, count: number = 1): Promise<any[]> {
```
Comment (Brooooooklyn) [inline packages/backend/server/src/cache/redis.ts]:
```suggestion
  async popBack<T = any>(key: string, count: number = 1): Promise<T[]> {
```

File: packages/backend/server/src/cache/redis.ts [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+  }
+
+  async list(key: string, start: number, end: number): Promise<any[]> {
+    return this.redis
+      .lrange(key, start, end)
+      .then(data => data.map(v => JSON.parse(v)))
+      .catch(() => []);
+  }
+
+  async popFront(key: string, count: number = 1): Promise<any[]> {
```
Comment (Brooooooklyn) [inline packages/backend/server/src/cache/redis.ts]:
```suggestion
  async popFront<T = any>(key: string, count: number = 1): Promise<T[]> {
```

File: packages/backend/server/src/cache/redis.ts [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+    return this.redis
+      .lpush(key, ...values.map(v => JSON.stringify(v)))
+      .catch(() => 0);
+  }
+
+  async len(key: string): Promise<number> {
+    return this.redis.llen(key).catch(() => 0);
+  }
+
+  async list(key: string, start: number, end: number): Promise<any[]> {
```
Comment (Brooooooklyn) [inline packages/backend/server/src/cache/redis.ts]:
```suggestion
  async list<T = any>(key: string, start: number, end: number): Promise<T[]> {
```

File: packages/backend/server/src/cache/redis.ts [hunk 6]
```diff
@@ -0,0 +1,186 @@
+import { Redis } from 'ioredis';
+
+import { Cache, CacheSetOptions } from './cache';
+
+export class RedisCache implements Cache {
+  constructor(private readonly redis: Redis) {}
+
+  // standard operation
+  async get(key: string): Promise<any> {
```
Comment (Brooooooklyn) [inline packages/backend/server/src/cache/redis.ts]:
```suggestion
  async get<T = unknown>(key: string): Promise<T | undefined> {
```

File: packages/backend/server/src/cache/cache.ts [hunk 7]
```diff
@@ -0,0 +1,314 @@
...
+
+  async len(key: string): Promise<number> {
+    return this.getArray(key).then(v => v?.value.length ?? 0);
+  }
+
+  /**
+   * list array elements with `[start, end]`
+   * the end indice is inclusive
+   */
+  async list(key: string, start: number, end: number): Promise<any[]> {
```
Comment (Brooooooklyn) [inline packages/backend/server/src/cache/cache.ts]:
```suggestion
  async list<T = unknown>(key: string, start: number, end: number): Promise<T[]> {
```

**Review by Brooooooklyn — CHANGES_REQUESTED** [review]
We need to add generics to the parts of the cache API that return the any type, with a default of unknown, to force the calling sites to provide the correct, detailed types explicitly.

File: packages/backend/server/src/cache/redis.ts [hunk 8]
```diff
@@ -0,0 +1,194 @@
+import { Redis } from 'ioredis';
```
Comment (darkskygit) [inline packages/backend/server/src/cache/redis.ts:1]: why dont use keyv's redis adapter? https://github.com/jaredwray/keyv/tree/main/packages/redis

Comment (forehalo) [reply packages/backend/server/src/cache/redis.ts:1]: it's a pure kv store abstraction, with only `get(k)`, `set(k, v)` interface, and we can't take advantages from redis's original list/hash structrue.

**Review by Brooooooklyn — APPROVED** [review]

**Pull request merged**